src/System.Management.Automation/engine/Utils.cs

```diff
@@ -1588,4 +1595,214 @@ public static void SetTestHook(string property, bool value)
+                    shellExecuteInfo.lpVerb = Marshal.StringToHGlobalUni(verb);
+                if (startInfo.Arguments.Length != 0)
+                    shellExecuteInfo.lpParameters = Marshal.StringToHGlobalUni(startInfo.Arguments);
+                if (startInfo.WorkingDirectory.Length != 0)
+                    shellExecuteInfo.lpDirectory = Marshal.StringToHGlobalUni(startInfo.WorkingDirectory);
+
+                shellExecuteInfo.fMask |= NativeMethods.SEE_MASK_FLAG_DDEWAIT;
+
+                if (!NativeMethods.ShellExecuteEx(shellExecuteInfo))
+                {
+                    int errorCode = Marshal.GetLastWin32Error();
+                    if (errorCode == 0) 
+                    {
+                        switch ((long)shellExecuteInfo.hInstApp)
+                        {
+                            case NativeMethods.SE_ERR_FNF: errorCode = NativeMethods.ERROR_FILE_NOT_FOUND; break;
+                            case NativeMethods.SE_ERR_PNF: errorCode = NativeMethods.ERROR_PATH_NOT_FOUND; break;
+                            case NativeMethods.SE_ERR_ACCESSDENIED: errorCode = NativeMethods.ERROR_ACCESS_DENIED; break;
+                            case NativeMethods.SE_ERR_OOM: errorCode = NativeMethods.ERROR_NOT_ENOUGH_MEMORY; break;
+                            case NativeMethods.SE_ERR_DDEFAIL:
+                            case NativeMethods.SE_ERR_DDEBUSY:
+                            case NativeMethods.SE_ERR_DDETIMEOUT: errorCode = NativeMethods.ERROR_DDE_FAIL; break;
+                            case NativeMethods.SE_ERR_SHARE: errorCode = NativeMethods.ERROR_SHARING_VIOLATION; break;
+                            case NativeMethods.SE_ERR_NOASSOC: errorCode = NativeMethods.ERROR_NO_ASSOCIATION; break;
+                            case NativeMethods.SE_ERR_DLLNOTFOUND: errorCode = NativeMethods.ERROR_DLL_NOT_FOUND; break;
+                            default: errorCode = (int)shellExecuteInfo.hInstApp; break;
+                        }
+                    }
+
+                    if(errorCode == NativeMethods.ERROR_BAD_EXE_FORMAT || errorCode == NativeMethods.ERROR_EXE_MACHINE_TYPE_MISMATCH)
+                    {
+                        throw new Win32Exception(errorCode, "InvalidApplication");
+                    }
+                    
+                    throw new Win32Exception(errorCode);
+                }
+            }
+            finally
+            {                
+                if (shellExecuteInfo.lpFile != (IntPtr)0) Marshal.FreeHGlobal(shellExecuteInfo.lpFile);
+                if (shellExecuteInfo.lpVerb != (IntPtr)0) Marshal.FreeHGlobal(shellExecuteInfo.lpVerb);
+                if (shellExecuteInfo.lpParameters != (IntPtr)0) Marshal.FreeHGlobal(shellExecuteInfo.lpParameters);
+                if (shellExecuteInfo.lpDirectory != (IntPtr)0) Marshal.FreeHGlobal(shellExecuteInfo.lpDirectory);
+            }
+
+            Process processToReturn = null;
```

Comment:
Could you please write a comment on what is happening right here? Please correct me if I am wrong. In the finally block you are freeing memory for lpfile, lpVerb, lpParameters, and lpDirectory. Then you use the pointer to the process to fetch the process using its id. My question is: Do we need to free anymore memory from shellExecuteInfo once the process is returned?